### PR TITLE
JDK-8315248: AssertionError in Name.compareTo

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a small fix to avoid an `AssertionError` coming out of `CharSequence.compare` for two `CharSequence` objects obtained from different instances of the `javax.lang.model` API.

Full details in the JBS issue and comments.

There are some minor cleanup edits as well, and some improvements to the code to throw AssertionError in a few places.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315248](https://bugs.openjdk.org/browse/JDK-8315248): AssertionError in Name.compareTo (**Bug** - P3)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15478/head:pull/15478` \
`$ git checkout pull/15478`

Update a local copy of the PR: \
`$ git checkout pull/15478` \
`$ git pull https://git.openjdk.org/jdk.git pull/15478/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15478`

View PR using the GUI difftool: \
`$ git pr show -t 15478`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15478.diff">https://git.openjdk.org/jdk/pull/15478.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15478#issuecomment-1698319715)

### CODE REVIEW

File: test/langtools/tools/javac/nametable/TestNameTables.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    Name.Table createNameTable(String option) {
+        Context c = new Context();
+        Options o = Options.instance(c);
+        o.put(option, "true");
+        Names n = new Names(c);
+        return n.table;
+    }
+
+    /**
+     * Tests operations using a single name tbale
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/nametable/TestNameTables.java:87]: typo: tbale

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Comment by archiecobbs** [comment]
This definitely works. However, if a class `X` implements `Comparable`, by definition instances must be able to `compareTo()` themselves with any other instance of type `X` (even if `X` is an interface - so this is a pretty major imposition that any class or interface implementing `Comparable` imposes on all subtypes, present, past, and future).

In this case, it is `Name` that implements `Comparable`, and `Utf8NameTable.NameImpl`'s version of `compareTo()` is just as an overriding optimization of `Name.compareTo()`, which does in fact work in all cases.

So it seems like the "truly correct" thing to do here would be something like this:
```java
// Utf8NameTable.NameImpl.compareTo()
@Override
public int compareTo(Name name0) {
    NameImpl name;
    try {
        name = (NameImpl)name0;
    } catch (ClassCastException e) {
        return super.compareTo(name0);
    }
    byte[] buf1 = getByteData();
    ...etc...
```
This probably doesn't matter in practice, but just wanted to point out the theoretical argument.

**Comment by jonathan-gibbons** [comment]
> This definitely works. However, if a class `X` implements `Comparable`, by definition instances must be able to `compareTo()` themselves with any other instance of type `X` (even if `X` is an interface - so this is a pretty major imposition that any class or interface implementing `Comparable` imposes on all subtypes, present, past, and future).
> 
> In this case, it is `Name` that implements `Comparable`, and `Utf8NameTable.NameImpl`'s version of `compareTo()` is just as an overriding optimization of `Name.compareTo()`, which does in fact work in all cases.
> 
> So it seems like the "truly correct" thing to do here would be something like this:
> 
> ```java
> // Utf8NameTable.NameImpl.compareTo()
> @Override
> public int compareTo(Name name0) {
>     NameImpl name;
>     try {
>         name = (NameImpl)name0;
>     } catch (ClassCastException e) {
>         return super.compareTo(name0);
>     }
>     byte[] buf1 = getByteData();
>     ...etc...
> ```
> 
> This probably doesn't matter in practice, but just wanted to point out the theoretical argument.

Noted; thanks. Since we're editing the code, I'll fix it.

**Comment by jonathan-gibbons** [comment]
>     NameImpl name;
>     try {
>         name = (NameImpl)name0;
>     } catch (ClassCastException e) {
>         return super.compareTo(name0);
>     }

See the updated code for a revised, more concise, suggestion using `instanceof` pattern.

**Pull request closed without merging**